nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/dao/ArchiveDao.java

```diff
@@ -0,0 +1,114 @@
+package com.nitorcreations.nflow.engine.internal.dao;
+
+import static com.nitorcreations.nflow.engine.internal.dao.DaoUtil.toTimestamp;
+import static com.nitorcreations.nflow.engine.internal.dao.DaoUtil.ColumnNamesExtractor.columnNamesExtractor;
+import static com.nitorcreations.nflow.engine.internal.storage.db.DatabaseConfiguration.NFLOW_DATABASE_INITIALIZER;
+import static org.apache.commons.lang3.StringUtils.join;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.joda.time.DateTime;
+import org.springframework.context.annotation.DependsOn;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.transaction.annotation.Transactional;
+
+import com.nitorcreations.nflow.engine.internal.config.NFlow;
+
+@Named
+@DependsOn(NFLOW_DATABASE_INITIALIZER)
+public class ArchiveDao {
+  private JdbcTemplate jdbc;
+  private TableMetadataChecker tableMetadataChecker;
+
+  @Inject
+  public void setJdbcTemplate(@NFlow JdbcTemplate jdbcTemplate) {
+    this.jdbc = jdbcTemplate;
+  }
+
+  @Inject
+  public void setTableMetadataChecker(TableMetadataChecker tableMetadataChecker) {
+    this.tableMetadataChecker = tableMetadataChecker;
+  }
+
+  public void ensureValidArchiveTablesExist() {
+    tableMetadataChecker.ensureCopyingPossible("nflow_workflow", "nflow_archive_workflow");
+    tableMetadataChecker.ensureCopyingPossible("nflow_workflow_action", "nflow_archive_workflow_action");
+    tableMetadataChecker.ensureCopyingPossible("nflow_workflow_state", "nflow_archive_workflow_state");
+  }
+
+  public List<Integer> listArchivableWorkflows(DateTime before, int maxRows) {
+    return jdbc.query(
+                    "select w.id id from nflow_workflow w, " +
+                    "(" +
+                    "  select parent.id from nflow_workflow parent " +
+                    "  where parent.next_activation is null and parent.modified <= ? " +
+                    "  and parent.root_workflow_id is null " +
+                    "  and not exists(" +
+                    "    select 1 from nflow_workflow child where child.root_workflow_id = parent.id " +
+                    "      and (child.modified > ? or child.next_activation is not null)" +
+                    "  )" +
+                    "  order by modified asc " +
+                    "  limit " + maxRows +
+                    ") as archivable_parent " +
+                    "where archivable_parent.id = w.id or archivable_parent.id = w.root_workflow_id",
+            new Object[] { toTimestamp(before), toTimestamp(before) }, new ArchivableWorkflowsRowMapper());
```

Comment:
swap arguments, rowmapper before arguments -> can get rid of the new Object[] (java compiler adds it automatically)